homeassistant/components/intesishome/climate.py

```diff
@@ -0,0 +1,364 @@
+
+import homeassistant.helpers.config_validation as cv
+
+from homeassistant.components import persistent_notification
+from homeassistant.components.climate import PLATFORM_SCHEMA, ClimateDevice
+from homeassistant.components.climate.const import (ATTR_HVAC_MODE,
+                                                    FAN_AUTO,
+                                                    FAN_HIGH,
+                                                    FAN_LOW,
+                                                    FAN_MEDIUM,
+                                                    HVAC_MODE_COOL,
+                                                    HVAC_MODE_DRY,
+                                                    HVAC_MODE_FAN_ONLY,
+                                                    HVAC_MODE_HEAT,
+                                                    HVAC_MODE_HEAT_COOL,
+                                                    HVAC_MODE_OFF,
+                                                    SUPPORT_FAN_MODE,
+                                                    SUPPORT_SWING_MODE,
+                                                    SUPPORT_TARGET_TEMPERATURE,
+                                                    SWING_BOTH,
+                                                    SWING_HORIZONTAL,
+                                                    SWING_OFF,
+                                                    SWING_VERTICAL)
+from homeassistant.const import (ATTR_TEMPERATURE, CONF_PASSWORD,
+                                 CONF_USERNAME, STATE_UNKNOWN,
+                                 TEMP_CELSIUS)
+from homeassistant.exceptions import PlatformNotReady
+
+REQUIREMENTS = ['pyintesishome==0.8']
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string
+})
+
+# Return cached results if last scan time was less than this value.
+# If a persistent connection is established for the controller, changes to
+# values are in realtime.
+SCAN_INTERVAL = timedelta(seconds=300)
+
+IH_FAN_QUIET = 'quiet'
+IH_SWING_WIDGET = 42
+
+MAP_OPERATION_MODE = {
+    'auto': HVAC_MODE_HEAT_COOL,
+    'cool': HVAC_MODE_COOL,
+    'dry': HVAC_MODE_DRY,
+    'fan': HVAC_MODE_FAN_ONLY,
+    'heat': HVAC_MODE_HEAT,
+    'off': HVAC_MODE_OFF
+}
+
+MAP_STATE_ICONS = {
+    HVAC_MODE_COOL: 'mdi:snowflake',
+    HVAC_MODE_DRY: 'mdi:water-off',
+    HVAC_MODE_FAN_ONLY: 'mdi:fan',
+    HVAC_MODE_HEAT: 'mdi:white-balance-sunny',
+    HVAC_MODE_HEAT_COOL: 'mdi:cached',
+}
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Create the IntesisHome climate devices."""
+    from pyintesishome import IntesisHome
+
+    ihuser = config[CONF_USERNAME]
+    ihpass = config[CONF_PASSWORD]
+
+    controller = IntesisHome(ihuser, ihpass, hass.loop)
+
+    await hass.async_add_executor_job(controller.connect)
```

Comment:
I've looked in the pyintesishome library and there are some issue with how asyncio is mixed with sync code.

`asyncio.ensure_future` is not threadsafe. This function is used here in the connect method to create the socket connection. I suggest instead making `IntesisHome.connect` a coroutine so we can use await inside it and await the creation of the socket.

`requests` is used to post to the api. I suggest to use the `aiohttp` library instead to be able to use asyncio for the post request, so it can be awaited. Have the library accept an optional aiohttp client session. Then we can provide the home assistant aiohttp client session to get best performance.

Reply:
Thanks very much - appreciate the advice, will work on that